server.go

```diff
@@ -1677,14 +1677,16 @@ func (s *Server) ListenAndServeTLSEmbed(addr string, certData, keyData []byte) e
 // the function will use previously added TLS configuration.
 func (s *Server) ServeTLS(ln net.Listener, certFile, keyFile string) error {
 	s.mu.Lock()
-	err := s.AppendCert(certFile, keyFile)
-	if err != nil && err != errNoCertOrKeyProvided {
-		s.mu.Unlock()
-		return err
-	}
-	if s.TLSConfig == nil {
-		s.mu.Unlock()
-		return errNoCertOrKeyProvided
+	s.configTLS()
+	config := s.TLSConfig.Clone()
+	var err error
+	configHasCert := len(config.Certificates) > 0 || config.GetCertificate != nil
+	if !configHasCert || certFile != "" || keyFile != "" {
```

Comment:
I would rather have it throw and error if it already has a certificate. That shows a mistake that we otherwise hide.

Reply:
```
fasthttp.ListenAndServeTLS()
fasthttp.ListenAndServe()
fasthttp.ListenAndServeTLSEmbed()
fasthttp.ListenAndServeUNIX()
```
The granularity of the above four fasthttp server usage methods is too large after all, and many server-side variables cannot be customized.
```
func main() {
	r := router.New()
	r.GET("/hello", func(ctx *fasthttp.RequestCtx) {
		ctx.Response.SetBody([]byte("hello workflow"))
	})
	s := &fasthttp.Server{
		Handler:   r.Handler,
		TLSConfig: &tls.Config{},
	}
	if err := s.ListenAndServeTLS("0.0.0.0:443", "", ""); err != nil {
		log.Println(err)
	}
}
```
Code as above ,If the user presets tls.Config in advance in this way, and does not write the decrypted SSL certificate key to the machine, then the user can only write the following code.
```
    s := &fasthttp.Server{
        Name: "FastHttp Server",
        Handler:   handler,
        ReadTimeout: readTimeout,
        WriteTimeout: writeTimeout,
        ReadBufferSize: maxHeaderBytes,
    }

    sslConfig := common.GetServerSSLConfig()
    if sslConfig.SSLEnabled {
        tlsConfig, err = GetServerTLSConfig(sslConfig.VerifyClient)
        if err != nil {
            return err
        }
        tcpLn, err := net.Listen("tcp4", addr)
        if err!=nil {
            return err
        }
        tlsLn := tls.NewListener(tcpLn, tlsConfig)
        return s.Serve(tlsLn)
    }
    return s.ListenAndServe(addr)
```

Therefore, I still hope to be consistent with the source package and do some compatibility. After all, in many development tasks, customizing server-side variables is a very common operation, but writing code to monitor the tls port, and other convenient The method is not very consistent. I still hope that fasthttp will package these details into the source code package

Reply:
```
func (s *Server) ServeTLSEmbed(ln net.Listener, certData, keyData []byte) error {
	s.mu.Lock()
	s.configTLS()
	configHasCert := len(s.TLSConfig.Certificates) > 0 || s.TLSConfig.GetCertificate != nil
	if configHasCert {
		return errAlreadyHasCertificate
	}

	if err := s.AppendCertEmbed(certData, keyData); err != nil {
		s.mu.Unlock()
		return err
	}

	// BuildNameToCertificate has been deprecated since 1.14.
	// But since we also support older versions we'll keep this here.
	s.TLSConfig.BuildNameToCertificate() //nolint:staticcheck

	s.mu.Unlock()

	return s.Serve(
		tls.NewListener(ln, s.TLSConfig.Clone()),
	)
}
```
If this is the case, then the custom fasthttp.Server calls ListenAndServeTLS. If certData and keyData are not provided, it will never be able to start with TLS. I understand that configHasCert should be compatible with providing certData and keyData to start TLS